File: 21-MergeTwoSortedLists/21-MergeTwoSortedLists.py
# Last updated: 9/16/2025, 12:10:48 AM
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        # If one of the lists is empty, return the other one
        if not list1:
            return list2
        if not list2:
            return list1

        # Ensure that list1 starts with the smaller element
        if list1.val > list2.val:
            list1, list2 = list2, list1

        # The head of the merged list is list1
        head = list1

        while list1 and list2:
            prev=None
            while list1 and list1.val<=list2.val:
                prev=list1
                list1=list1.next
            prev.next=list2
            list1,list2=list2,list1

        return head        

File: 21-MergeTwoSortedLists/21-MergeTwoSortedLists.cpp
// Last updated: 9/16/2025, 12:10:44 AM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* a, ListNode* b) {
        ListNode* dummy=new ListNode(0);
        ListNode* curr=dummy;

        while(a && b){
            if(a->val<b->val){
                curr->next=a;
                a=a->next;
            }else{
                curr->next=b;
                b=b->next;
            }
            curr=curr->next;
        }

    if (a) curr->next = a;
    else curr->next = b;

    return dummy->next;

    }
};